[FLINK-21188] Replaces ArchivedExecutionGraph by ExecutionGraphInfo

XComp opened the pull request:
## What is the purpose of the change

We want to forward the exception history (see [FLINK-6042](https://issues.apache.org/jira/browse/FLINK-6042)) as part of a wrapper class. This enables us to extend the objects passed forward to the history server and ExecutionGraph processing REST handlers more easily (considering possible changes here in the context of the declarative scheduler work in [FLIP-160](https://cwiki.apache.org/confluence/display/FLINK/FLIP-160%3A+Declarative+Scheduler)).

## Brief change log

* Initially, I introduced `WithExecutionGraph` and `WithExceptionHistory` as separate interfaces. I would oppose this decision now due to the way the `ExecutionGraphInfo` is used.
* `ExecutionGraphInfo` replaces almost all occurrences of `ArchivedExecutionGraph`.
* `AbstractArchivedExecutionGraphHandler` is introduced to have all subclasses of the old `AbstractExecutionGraphHandler` (except for `JobExceptionsHandler`) use this new abstract parent class, instead.
* The `JsonArchivist` was extended by a default implementation handling `ExecutionGraphInfo`.

## Verifying this change

No functionality is changed. All present test should still pass.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherGateway.java:
```diff
@@ -73,7 +74,7 @@
      */
     @Override
-    CompletableFuture<ArchivedExecutionGraph> requestJob(JobID jobId, @RpcTimeout Time timeout);
+    CompletableFuture<ExecutionGraphInfo> requestJob(JobID jobId, @RpcTimeout Time timeout);
```
It might be a bit easier to not change the return type of this method and instead to introduce a new method `CompletableFuture<ExecutionGraphInfo> requestJobInfo()`. Then, one could even add a default implementation of this method to call `requestJobInfo().thenApply(ExecutionGraphInfo::getArchivedExecutionGraph)`.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @XComp.

I think your proposed changes could be a bit more targeted towards the rest handlers. Try to think along the lines that you can also introduce a new method `CompletableFuture<ExecutionGraphInfo> requestJobInfo` to the `DispatcherGateway`. That way your change won't probably affect the whole stack from the client to the cluster.

Ideally, only the rest handlers, the `ExecutionGraphCache` and the `ArchivedExecutionGraphStore` are affected by these changes.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/OnCompletionActions.java:
```diff
@@ -18,18 +18,19 @@
      * Job reached a globally terminal state.
      *
-     * @param executionGraph serializable execution graph
+     * @param executionGraphInfo serializable {@link AccessExecutionGraph}-related information
```
```suggestion
     * @param executionGraphInfo executionGraphInfo contains information about the terminated job
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerImpl.java:
```diff
@@ -241,10 +241,10 @@ public void start() throws Exception {
-    public void jobReachedGloballyTerminalState(ArchivedExecutionGraph executionGraph) {
+    public void jobReachedGloballyTerminalState(ExecutionGraphInfo executionGraphInfo) {
         unregisterJobFromHighAvailability();
         // complete the result future with the terminal execution graph
```
Comment seems to be outdated with this change.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerResult.java:
```diff
@@ -56,9 +57,10 @@ public boolean isInitializationFailure() {
      * @throws IllegalStateException if the result is not a success
      */
-    public ArchivedExecutionGraph getArchivedExecutionGraph() {
+    // TODO: rename method and javadoc
```
I guess you still wanna do this, right?

XComp replied:
Good that you found that. It appears that I did something wrong when doing a final lookup for TODOs

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerResult.java:
```diff
@@ -81,21 +83,23 @@ public boolean equals(Object o) {
         }
         JobManagerRunnerResult that = (JobManagerRunnerResult) o;
-        return Objects.equals(archivedExecutionGraph, that.archivedExecutionGraph)
+        // TODO: verify ExecutionGraphInfo.equals
```
still valid?

XComp replied:
I wanted to re-iterate over the `equals`/`hashCode` contract for `ExecutionGraphInfo` and `ArchivedExecutionGraph`. The `JobManagerRunnerResult` must have slipped through when going through the `TODO`s.

I was just wondering whether there should have been an `equals`/`hashCode` implementation for `ArchivedExecutionGraph` in the first place. Isn't `ArchivedExecutionGraph` well-defined by its `ArchivedExecutionJobVertex` instances stored in `ArchivedExecutionGraph::tasks`? And `ArchivedExecutionJobVertex` would be well-defined through its `ArchivedExecutionVertex` instances stored in `ArchivedExecutionJobVertex::taskVertices` which would be defined through its `ArchivedExecutionVertex::currentExecution`? I guess, we wouldn't gain anything code-wise besides having the contract formulated in the code which might increase readability. Any thoughts on this @zentol @tillrohrmann ?

tillrohrmann replied:
I think `JobManagerRunnerResult` should not have an `equals` and `hashCode` method. Usually only objects which you want to compare or use in a `HashMap` should implement these methods. We sometimes added these methods because we did not write a `Matcher` when using it in tests. However, this was more because of laziness than good software engineering. The problem is that if this is only used in tests, then it should probably be a test concern.

XComp replied:
I see, makes sense.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerRunnerResult.java:
```diff
@@ -81,21 +83,23 @@ public boolean equals(Object o) {
     @Override
     public int hashCode() {
-        return Objects.hash(archivedExecutionGraph, failure);
+        // TODO: verify ExecutionGraphInfo.hashCode
```
still valid?

XComp replied:
See [related discussion](https://github.com/apache/flink/pull/14804#discussion_r580067125) for the corresponding `equals()` method.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/AbstractArchivedExecutionGraphHandler.java:
```diff
@@ -0,0 +1,74 @@
+ * @param <R> the response type
+ * @param <M> the message parameter type
+ */
+public abstract class AbstractArchivedExecutionGraphHandler<
```
Maybe call it `AbstractAccessExecutionGraphHandler` since we are giving access to the `AccessExecutionGraphHandler`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java:
```diff
@@ -530,20 +531,22 @@ public void testCacheJobExecutionResult() throws Exception {
-                        .setState(expectedState)
-                        .setFailureCause(new ErrorInfo(new RuntimeException("expected"), 1L))
-                        .build();
+        final ExecutionGraphInfo failedExecutionGraph =
```
```suggestion
        final ExecutionGraphInfo failedExecutionGraphInfo =
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/FileExecutionGraphInfoStoreTest.java:
```diff
@@ -220,40 +226,46 @@ public void testCloseCleansUp() throws IOException {
+                    executionGraphStore.getExecutionGraphInfoCache();
 
-            Collection<ArchivedExecutionGraph> executionGraphs = new ArrayList<>(64);
+            Collection<ExecutionGraphInfo> executionGraphs = new ArrayList<>(64);
```
```suggestion
            Collection<ExecutionGraphInfo> executionGraphInfos = new ArrayList<>(64);
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/FileExecutionGraphInfoStoreTest.java:
```diff
@@ -267,18 +279,18 @@ public void testCacheLoading() throws IOException {
                     storageDirectory.listFiles().length, Matchers.equalTo(executionGraphs.size()));
 
-            for (ArchivedExecutionGraph executionGraph : executionGraphs) {
+            for (ExecutionGraphInfo executionGraph : executionGraphs) {
```
```suggestion
            for (ExecutionGraphInfo executionGraphInfo : executionGraphInfos) {
```

tillrohrmann reviewed (approved):
Thanks for updating this PR @XComp. The changes look good to me. It is still not few changes but I think we won't get around what this PR includes. I had a few minor comments. Once we have FLINK-21187 merged, we should be able to proceed with merging this PR.

tillrohrmann reviewed (approved):
Thanks for addressing my comments @XComp. LGTM. Can I merge the PR as is?

XComp commented:
I squashed and reorganized the commits. We can proceed with merging it after the final Azure build is successful.

XComp commented:
I had to give it another run because I missed an unused import during - I should solely rely on Intellij for rebasing. 🙄 I fixed the issues and rebased the branch once more.

tillrohrmann commented:
Alrighty. AZP has passed. Merging this PR now.
